### ISSUE TITLE
Error launching source server: json: cannot unmarshal array into Go value of type string

### ISSUE THREAD

**Issue by jcejohnson** [hashicorp/packer#4990]
- Packer version from `packer version`: 0.12.3 and 1.0.0
- Host platform: Mac & Linux
- Debug log output from `PACKER_LOG=1 packer build template.json`.
- The _simplest example template and scripts_ needed to reproduce the bug.
  ~~Sadly, our corporate policies do not give us access to gist.github.com but I've attached a packer.log and packer.txt here.~~ Fixed

Using the openstack builder with `source_image_name` to identify the source, I have found that if any image has a property whose value is an array, packer will fail with the error:
    Error launching source server: json: cannot unmarshal array into Go value of type string

For instance:
```
    openstack image set --property block_device_mapping='[{"guest_format": null, "boot_index": 0,  "delete_on_termination": true, "no_device": null, "snapshot_id": "3222ae9c-e79c-4d65-adab-132effa47db4", "device_name": "/dev/vda", "disk_bus": "virtio", "image_id": null, "source_type": "snapshot", "device_type": "disk", "volume_id": null, "destination_type": "volume", "volume_size": 10}]' some-unrelated-image
```

That particular property value was taken from an image that was created by snapshotting a VM.

```json
{
    "builders": [
        {
            "domain_name": "{{ user `OS_DOMAIN_NAME` }}",
            "flavor": "kujo.base",
            "floating_ip_pool": "{{ user `OS_FLOATING_IP_POOL` }}",
            "identity_endpoint": "{{ user `OS_AUTH_URL` }}",
            "image_name": "base-test-build-1.2.12",
            "insecure": "true",
            "name": "base-test-build",
            "networks": [
                "{{ user `OS_NETWORK_UUID` }}"
            ],
            "password": "{{ user `OS_PASSWORD` }}",
            "region": "{{ user `OS_REGION_NAME` }}",
            "security_groups": "{{ user `OS_SECURITY_GROUP` }}",
            "source_image_name": "minimal-kujo-centos-6-1.2.12",
            "ssh_disable_agent": "true",
            "ssh_keypair_name": "{{ user `OS_KEY_NAME` }}",
            "ssh_private_key_file": "{{ user `SECRETS` }}/kujo/insecure-key",
            "ssh_username": "root",
            "tenant_name": "{{ user `OS_TENANT_NAME` }}",
            "type": "openstack",
            "username": "{{ user `OS_USERNAME` }}"
        }
    ],
    "variables": {
        "OS_AUTH_HOST": "{{ env `OS_AUTH_HOST` }}",
        "OS_AUTH_PORT": "{{ env `OS_AUTH_PORT` }}",
        "OS_AUTH_PROTO": "{{ env `OS_AUTH_PROTO` }}",
        "OS_AUTH_TOKENS_URL": "{{ env `OS_AUTH_TOKENS_URL` }}",
        "OS_AUTH_URL": "{{ env `OS_AUTH_URL` }}",
        "OS_AUTH_URL_v2": "{{ env `OS_AUTH_URL_v2` }}",
        "OS_AUTH_URL_v3": "{{ env `OS_AUTH_URL_v3` }}",
        "OS_CACERT": "{{ env `OS_CACERT` }}",
        "OS_DESCRIPTION": "{{ env `OS_DESCRIPTION` }}",
        "OS_DOMAIN_NAME": "{{ env `OS_DOMAIN_NAME` }}",
        "OS_ENDPOINT_TYPE": "{{ env `OS_ENDPOINT_TYPE` }}",
        "OS_FLAVOR_REF": "{{ env `OS_FLAVOR_REF` }}",
        "OS_FLAVOR_UUID": "{{ env `OS_FLAVOR_UUID` }}",
        "OS_FLOATING_IP_MAX": "{{ env `OS_FLOATING_IP_MAX` }}",
        "OS_FLOATING_IP_MIN": "{{ env `OS_FLOATING_IP_MIN` }}",
        "OS_FLOATING_IP_POOL": "{{ env `OS_FLOATING_IP_POOL` }}",
        "OS_HELION_MAJOR": "{{ env `OS_HELION_MAJOR` }}",
        "OS_IDENTITY_API_VERSION": "{{ env `OS_IDENTITY_API_VERSION` }}",
        "OS_IMAGE_REF_CENTOS_6": "{{ env `OS_IMAGE_REF_CENTOS_6` }}",
        "OS_IMAGE_REF_CENTOS_7": "{{ env `OS_IMAGE_REF_CENTOS_7` }}",
        "OS_IMAGE_REF_RHEL_6": "{{ env `OS_IMAGE_REF_RHEL_6` }}",
        "OS_IMPLEMENTATION": "{{ env `OS_IMPLEMENTATION` }}",
        "OS_INTERFACE": "{{ env `OS_INTERFACE` }}",
        "OS_KEY_NAME": "{{ env `OS_KEY_NAME` }}",
        "OS_NETWORK_REF": "{{ env `OS_NETWORK_REF` }}",
        "OS_NETWORK_UUID": "{{ env `OS_NETWORK_UUID` }}",
        "OS_PASSWORD": "{{ env `OS_PASSWORD` }}",
        "OS_PRIV_KEY": "{{ env `OS_PRIV_KEY` }}",
        "OS_PROJECT_DOMAIN_NAME": "{{ env `OS_PROJECT_DOMAIN_NAME` }}",
        "OS_PROJECT_NAME": "{{ env `OS_PROJECT_NAME` }}",
        "OS_REGION_NAME": "{{ env `OS_REGION_NAME` }}",
        "OS_SECURITY_GROUP": "{{ env `OS_SECURITY_GROUP` }}",
        "OS_TENANT_NAME": "{{ env `OS_TENANT_NAME` }}",
        "OS_USERNAME": "{{ env `OS_USERNAME` }}",
        "OS_USER_DOMAIN_NAME": "{{ env `OS_USER_DOMAIN_NAME` }}",
        "OS_VM_USERNAME": "{{ env `OS_VM_USERNAME` }}",
        "SECRETS": "{{ env `SECRETS` }}",
        "WINRM_PASSWORD": "{{ env `WINRM_PASSWORD` }}"
    }
}
```

Packer log see [gist](https://gist.github.com/rickard-von-essen/1706d6de18af52a1932eef68a0a299ff)

**Comment by rickard-von-essen** [comment]
Sorry I don't understand? What do you try to set explicitly?

**Comment by rickard-von-essen** [comment]
Could you include the full command you use to execute Packer with.

**Comment by jcejohnson** [comment]
The packer command is simply: packer build packer.json.

If I use "openstack image set --property something='[ ... ]'" on any image the build will fail with the unmarshal error. As it turns out, openstack itself does this behind the scenes when creating a snapshot image (i.e. -- "openstack server image create <servername> --name <snapshot-name>"). So I can't simply tell people "don't do that" since openstack itself is doing it.

**Comment by rickard-von-essen** [comment]
Ok, thanks for the explanation. The way to repro this:

- `openstack image set --property something='["a", "b"]' <source-image>`
- Run `packer build template.json` with `"source_image": "<source-image>"`

**Comment by rickard-von-essen** [comment]
This fails on creating the source instance at [step_run_source_server.go - L83](https://github.com/hashicorp/packer/blob/master/builder/openstack/step_run_source_server.go#L83). Most likely an upstream bug.

**Comment by jcejohnson** [comment]
Another interesting detail for anyone who may run across this.

The offending image in question has a block_device_mapping property that contains array data. That image was created via 'openstack server image create <my-vm>'.  <my-vm> was created via 'openstack server create --volume <some-volume> ... <my-vm>'.

Images created from VMs that were, in turn, created from *images* do not have this problem. Specifically, they do not have the block_device_mapping property because there is no volume involved.

TL;DR - If you're snapshotting volume-backed VMs those snapshots will have a block_device_mapping that will cause this error.

**Comment by jdanekrh** [comment]
I believe this may have same cause as previously fixed https://github.com/hashicorp/terraform/issues/10767. The fix was in upgrading version of the Gophercloud library.

Anyways, it might be good to align version of Gophercloud between packer, terraform, and any additional product in this family where applicable.

**Comment by rickard-von-essen** [comment]
Could you please reference the Gophercloud version that fixes this.

**Comment by jdanekrh** [comment]
I have no idea. Something between what terraform/packer had when their respective issues were reported, and what they have now.

There was an upgrade of gopherclouds for packer about a month ago I've just confirmed that build of git master of packer is not affected. Version 1.0.0 fails for me, but the master works for me fine, on the same .json config, against the same OpenStack server.

**Comment by jdanekrh** [comment]
@jcejohnson can you retest with latest release, or with master? Current master worked for me, and the fix (dependency upgrade) should be in 1.0.3.

**Comment by trodemaster** [comment]
I was able to do a quick test this morning. Set one of my packer templates to source a volume backed snapshot and added  something ["a", "b"] as a custom property. I did not observe any error like I previously was getting. Packer v1.0.3-dev (9bc0e17e)

Thanks,
Blake

**Comment by jcejohnson** [comment]
Yea, I can retest it in the next day or two. We changed our process to avoid the issue so I'll have to contrive a scenario & it'll be easier to focus on that when I'm not in the office.

**Comment by jcejohnson** [comment]
Confirmed. Version 1.0.3 fixes this issue. Thanks!
